src/sinks/vector/v2.rs

```diff
@@ -0,0 +1,312 @@
+                    .ca_certificate(ca)
+                    .domain_name(host);
+
+                endpoint.tls_config(tls_config)?
+            }
+            None => endpoint,
+        };
+
+        let client = proto::Client::new(endpoint.connect_lazy()?);
+        let healthcheck = healthcheck(client.clone(), cx.healthcheck.clone());
+
+        let request = self.request.unwrap_with(&REQUEST_DEFAULTS);
+        let batch = BatchSettings::default()
+            .bytes(1300)
+            .events(1000)
+            .timeout(1)
+            .parse_config(self.batch)?;
+
+        let svc = ServiceBuilder::new()
+            .settings(request, VectorGrpcRetryLogic)
+            .service(client);
+
+        let buffer = VecBuffer::new(batch.size);
+        let sink = BatchSink::new(svc, buffer, batch.timeout, cx.acker())
+            .sink_map_err(|error| error!(message = "Fatal Vector GRPC sink error.", %error))
+            .with_flat_map(move |event| stream::iter(encode_event(event)).map(Ok));
+
+        Ok((VectorSink::Sink(Box::new(sink)), Box::pin(healthcheck)))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Any
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "vector_grpc"
+    }
+}
+
+/// Check to see if the remote service accepts new events.
+async fn healthcheck(mut client: Client, options: SinkHealthcheckOptions) -> crate::Result<()> {
+    // TODO: if enabled, allow changing the endpoint through `options.uri`.
+    if !options.enabled {
+        return Ok(());
+    }
+
+    let request = client.health_check(proto::HealthCheckRequest {});
+
+    if let Ok(response) = request.await {
+        let status = proto::ServingStatus::from_i32(response.into_inner().status);
+
+        if let Some(proto::ServingStatus::Serving) = status {
+            return Ok(());
+        }
+    }
+
+    Err(Box::new(Error::Health))
+}
+
+fn get_authority(url: &str) -> Result<String, Error> {
+    url.parse::<Uri>()
+        .ok()
+        .and_then(|uri| uri.authority().map(ToString::to_string))
+        .ok_or(Error::NoHost)
+}
+
+impl tower::Service<Vec<proto::EventRequest>> for Client {
+    type Response = ();
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Readiness check of the client is done through the `push_events()`
+        // call happening inside `call()`. That check blocks until the client is
+        // ready to perform another request.
+        //
+        // See: <https://docs.rs/tonic/0.4.2/tonic/client/struct.Grpc.html#method.ready>
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, requests: Vec<proto::EventRequest>) -> Self::Future {
+        let mut futures = Vec::with_capacity(requests.len());
+
+        // TODO: Instead of firing off multiple requests, have the server accept
+        // more than one event per request (i.e. bulk endpoint).
```

Comment:
I'll bench this, and probably do the change in a follow-up PR. I suspect batching will be faster, but this might already be "good enough" to get this PR over the finish line.

Reply:
FWIW batching is almost always going to play well on modern machines, networks.

Reply:
Yeah I'd expect batching to make a very significant difference here. We can bench it, but there's almost zero chance we don't want it.

Reply:
This is going to add extra overhead for end-to-end acknowledgements as well, as it will take extra effort to track all of the acknowledgements separately.